File: Server/Models/Event.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace WebApplication.Models
{
    [JsonConverter(typeof(EventConverter))]
    public class Event
    {
        public EventType type { get; set; }
        public dynamic data { get; set; }

        public Event(EventType type, dynamic data)
        {
            this.type = type;
            this.data = data;
        }
    }

    [JsonConverter(typeof(StringEnumConverter))]
    public enum EventType
    {
        NEW_MESSAGE, MESSAGE_RELAY, AUTHENTICATION, AUTHENTICATION_RESULT, REGISTER, REGISTER_RESULT, MESSAGE_FETCH, MESSAGE_RETURN
    }

    public class RegisterResult
    {
        public bool success { get; set; }
        public string reason { get; set; }

        public RegisterResult(bool success, string reason)
        {
            this.success = success;
            this.reason = reason;
        }

        public RegisterResult(bool success)
        {
            this.success = success;
        }
    }

    public class Autentication
    {
        public string username { get; set; }

        public string password { get; set; }
    }
}

File: Proxy/wwwroot/app/Event.ts

export class Event {
    type: string;
    data: any;

    public constructor(type: string, data: any) {
        this.type = type;
        this.data = data;
    }
}

export class Authentication {
    username: string;
    password: string;

    public constructor(username: string, password: string) {
        this.username = username;
        this.password = password;
    }
}

export class RegisterResult {
    success: boolean;
    reason: string;

    public constructor(success: boolean, reason: string) {
        this.success = success;
        this.reason = reason;
    }
}